### ISSUE TITLE
Support for custom types

### ISSUE THREAD

**Issue by liveforeverx** [bitwalker/conform#27]
It would be very nice to have validation types.

Example:
```elixir
  datatype: Module
```

Where `Module` implements behaviour, implements function `validate`, which returns {:ok, value} | {:error, error}

It will allow to build own validation libraries on top of conform.

**Comment by bitwalker** [comment]
I like this! Would you be open to implementing this? I unfortunately am suuuuper constrained on time at the moment, and I would like to keep things moving along. I'll certainly merge it if it's feature complete.

**Comment by liveforeverx** [comment]
Yeah, I think, that I will implement, and check #9 in this case too.

We have used conform for various releases, and have massive code duplication at the moment, which should be get rided with this 2 issues.

**Comment by bitwalker** [comment]
If I can be of any help, let me know, I'd be glad to contribute what I can.

**Comment by 0xAX** [comment]
Hello @bitwalker, I'm working on this issue for now and what do you think about following implementation:

An user will provide schema that will contain mapping with a module as custom data type, like this:

```elixir
    "myapp.Custom.Enum": [
      doc: "Provide documentation for myapp.Custom.Enum here.",
      to: "myapp.Custom.Enum",
      datatype: [{MyModule, [:dev, :prod, :test]}],
      default: :dev
     ]
```

The `Mymodule` may provide following API:

  * `to_doc/1` - will append doc string for the given mapping;
  * `transition/2` and `transition/3` - will provide default transition for the given mapping if it does not exists in the schema;
  * `parse_datatype/2` - will provide ability to cast value of the given mapping to the certain type.

For example we are going to add custom enum to the conform base as `Conform.Type.Enum` module to show user how a custom type can be built. It will have the same behavior as internal enum. The `to_doc` of the `Conform.Type.Enum` will append documentation:

```elixir
def to_doc(values) do
  "Allowed values: #{Enum.join(values, ", ")}\n"
end
```

The transition will validate and execute transition of the given value:

```elixir
def transition(mapping, val) do
  check_enum(mapping, val)
end

defp check_enum(mapping, val) do
  valid_values = Keyword.get(mapping, :datatype) |> Keyword.get(:'Elixir.Conform.Type.Enum')
  case Keyword.get(mapping, :default) do
    nil ->
      val
    _default ->
      if val in valid_values do
        val
      else
        {:error, "Wrong value of the enum"}
      end
  end
end
```

And the `parse_datatype` will convert config value to atom (it is only for enum, user will be able to provide own type casting):

```elixir
  def parse_datatype(_key, val) when is_list(val) do
    List.to_atom(val)
  end
```


What do you think about this? 

cc: @liveforeverx

**Comment by 0xAX** [comment]
@bitwalker I've pushed this implementation in #36. What do you think about it?

**Comment by bitwalker** [comment]
Looks pretty much ready to merge, just one comment on the README and we're good to go!

**Comment by bitwalker** [comment]
Merged via #36, will be pushed in a new release today.
